src/Stratis.Bitcoin/Consensus/ChainedHeaderTree.cs

```diff
@@ -216,35 +214,55 @@ public List<int> PartialOrFullValidationFailed(ChainedHeader chainedHeader)
         }
 
         /// <summary>
-        /// Recursively remove all the branches in the tree that are after the given <see cref="chainedHeader"/> (including) and return all the peers that where discovered.
+        /// Remove all the branches in the tree that are after the given <see cref="startHeader" /> (including)
+        /// and return all the peers that where claiming next headers.
         /// </summary>
-        /// <remarks>
-        /// The <see cref="discoveredPeers"/> will be modified in the processes.
-        /// </remarks>
-        /// <param name="discoveredPeers">List of peers that where discovered.</param>
-        /// <param name="chainedHeader">The chained header to start from.</param>
-        private void RemoveNextClaimsRecursive(List<int> discoveredPeers, ChainedHeader chainedHeader)
+        /// <param name="startHeader">The chained header to start from.</param>
+        /// <returns>List of peer Ids for banning.</returns>
+        private List<int> RemoveNextClaims(ChainedHeader startHeader)
         {
-            this.logger.LogTrace("({0}:'{1}',{2}.{3}:{4})", nameof(chainedHeader), chainedHeader, nameof(discoveredPeers), nameof(discoveredPeers.Count), discoveredPeers.Count);
+            this.logger.LogTrace("({0}:'{1}')", nameof(startHeader), startHeader);
 
-            foreach (ChainedHeader nextChainedHeader in chainedHeader.Next)
-            {
-                this.RemoveNextClaimsRecursive(discoveredPeers, nextChainedHeader);
-            }
-            
-            this.chainedHeadersByHash.Remove(chainedHeader.HashBlock);
-            chainedHeader.Previous.Next.Remove(chainedHeader);
+            var peersToBan = new List<int>();
 
-            if (this.peerIdsByTipHash.TryGetValue(chainedHeader.HashBlock, out HashSet<int> peers))
+            var headersToProcess = new Stack<ChainedHeader>();
+            headersToProcess.Push(startHeader);
+
+            while (headersToProcess.Count != 0)
             {
-                foreach (int peer in peers)
+                ChainedHeader header = headersToProcess.Pop();
+
+                foreach (ChainedHeader nextHeader in header.Next)
+                    headersToProcess.Push(nextHeader);
+                
+                if (this.peerIdsByTipHash.TryGetValue(header.HashBlock, out HashSet<int> peers))
                 {
-                    this.peerTipsByPeerId.Remove(peer);
+                    foreach (int peerId in peers)
+                    {
+                        if (peerId != LocalPeerId)
```

Comment:
If a peer failed validation we probably want to remove it even if its claimed by us no?

Reply:
We will claim something that is not fully validated only during a reorg. In case reorg fails we will rewind to prev tip and remove the claim. It will be done by the ConsensusManager